NTC with multi time series : internal structures

guilpier-code opened the pull request.

guilpier-code committed:
[skip ci] NTC multi TS - load and save : first developments (does not compile)

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -125,19 +129,23 @@ void AreaLink::reverse()
     // inverting NTC values
-    double* tmp = new double[data.height];
-    size_t siz = sizeof(double) * data.height;
+    double* tmp = new double[directCapacities.height];
+    size_t size = sizeof(double) * directCapacities.height;
 
-    memcpy(tmp, data[fhlNTCDirect], siz);
-    memcpy(data[fhlNTCDirect], data[fhlNTCIndirect], siz);
-    memcpy(data[fhlNTCIndirect], tmp, siz);
+    memcpy(tmp, directCapacities[0], size);
+    memcpy(directCapacities[0], indirectCapacities[0], size);
+    memcpy(indirectCapacities[0], tmp, size);
     delete tmp;
 
-    data.markAsModified();
-    data.flush();
+    directCapacities.markAsModified();
+    indirectCapacities.markAsModified();
+
+    directCapacities.flush();
+    indirectCapacities.flush();
```
Why do we need to swap memory content with deep copies ? Why not just swap pointers here ?

flomnes replied:
See #439

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -336,38 +344,58 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
                   && ret;
-            link.data.pasteToColumn(fhlNTCDirect, tmp[0]);
+            link.directCapacities.pasteToColumn(0, tmp[0]);
+            link.directCapacities.markAsModified();
```
Matrix `link.directCapacities` only contains link data. Therefore there is no need for a temporary matrix here.

guilpier-code replied:
ok, done

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -377,34 +405,32 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
-                link.data[fhlHurdlesCostDirect][i] = temp[fhlHurdlesCostDirect][i];
-                link.data[fhlHurdlesCostIndirect][i] = temp[fhlHurdlesCostIndirect][i];
+                link.directCapacities[0][i] = tmpMatrix[0][i];
+                link.indirectCapacities[0][i] = tmpMatrix[1][i];
```
Is this piece of code responsible for loading old studies ?  It cannot load new studies with more than NTC timeseries.

guilpier-code replied:
Yes, here we load old studies (between 320 and 630), where NTC have a single timeseries per link

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -336,38 +344,58 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
         if (study.header.version < 320)
         {
-            link.data.resize(fhlMax, HOURS_PER_YEAR);
-            link.data.zero();
-            if (link.data.jit)
+            // gp : refactor this part : avoid code duplication
+            link.parameters.resize(fhlMax, HOURS_PER_YEAR);
```
Consider a new function for old studies. This function is way too long

guilpier-code replied:
A possible solution could be something like : 
```c++
linkLoaderInterface* linkLoader; // Imagine a smart pointer here, not a regular pointer

// A factory to build the right link loader
if (study.header.version < 320)
    linkLoader = new linkLoader_under320(...)
else if (study.header.version < 630)
    linkLoader = new linkLoader_320_to_630(...)
else if (study.header.version < 820)
   linkLoader = new linkLoader_630_to_820(...)
else
  linkLoader = new linkLoader_820(...)

linkLoader.run();
```
To be discussed...

guilpier-code replied:
I extracted functions instead

guilpier-code committed:
NTC multi TS - load and save : splitting link's data into 3 matrices
- 1 column matrices for direct/indirect capacities
- 6 columns matrices for the other TS parameters

guilpier-code committed:
[skip ci] NTC multi TS - load and save : fix an assertion issue when loading links TS

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/ntc-multi-ts--load-and-save

guilpier-code committed:
NTC multi TS - split link's data : fix compilation problem after previous merge

Merge from another branch brought new tests that make an old usage of links data.
So the tests source file did not compile anymore.
Current commit solves this.

flomnes committed:
Introduce and use Matrix<>::swap (#439)

* Introduce and use Matrix<>::swap

* Remove template arguments (PR remarks)

* Add possibly missing header

* Add keyword noexcept to Matrix<>::swap

flomnes commented on src/libs/antares/study/area/links.h:
```diff
@@ -100,6 +100,10 @@ class AreaLink final : public Yuni::NonCopyable<AreaLink>
     void markAsModified() const;
+
+    void setPathToDataFile(const AnyString& path) { pathToDataFile = path; }
+    
+    bool loadDataFromCVSfile(uint loadOptions);
```
~~CVS~~ CSV

guilpier-code replied:
done

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -337,38 +346,58 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
             Matrix<double> tmp;
-            // NTC
+            // NTC direct
             buffer.clear() << folder << SEP << link.with->id << SEP << "direct."
                            << study.inputExtension;
```
Check spec to use the right path

guilpier-code replied:
what do you mean ? (this is the path for direct capacities data file of a very old versions of the study)

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -337,38 +346,58 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
             ret = tmp.loadFromCSVFile(
-                    buffer, 1, 8760, Matrix<>::optFixedSize | Matrix<>::optImmediate)
+                    buffer, 1, HOURS_PER_YEAR, Matrix<>::optFixedSize | Matrix<>::optImmediate)
                   && ret;
-            link.data.pasteToColumn(fhlNTCDirect, tmp[0]);
+            link.directCapacities.pasteToColumn(0, tmp[0]);
```
There is no need to use a buffer. Load the file contents into `directMatrix` directly.

It wasn't possible previously, but now it is.

guilpier-code replied:
ok

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -337,38 +346,58 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
+                    buffer, 1, HOURS_PER_YEAR, Matrix<>::optFixedSize | Matrix<>::optImmediate)
                   && ret;
-            link.data.pasteToColumn(fhlNTCIndirect, tmp[0]);
+            link.indirectCapacities.pasteToColumn(0, tmp[0]);
+            link.indirectCapacities.markAsModified();
 
```
Same as above: no need for a tmp buffer.

guilpier-code replied:
ok

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -378,136 +407,95 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
-            ret = link.data.loadFromCSVFile(buffer, fhlMax, HOURS_PER_YEAR, Matrix<>::optFixedSize)
-                  && ret;
+            link.setPathToDataFile(buffer);
+            ret = link.loadDataFromCVSfile(Matrix<>::optFixedSize | Matrix<>::optImmediate) && ret;
```
Why do we need to add flag `Matrix<>::optImmediate` ?

guilpier-code replied:
There's a problem otherwise : if we don't add this load option, the data is not loaded and code waits until it's the right time (jit : just in time) to load it (that is when we really use it). 
Because we want to update the new data structure (data is now split in 3 parts) now, we need to really load the old 8 columns matrix.
The way to update the new data structure without actually load the data might be possible, but at this stage of development, I keep it simple and load old data to construct the new one.

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -591,23 +579,25 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
 
         // memory swap
-        link.data.flush();
+        link.parameters.flush();
+        link.directCapacities.flush();
+        link.indirectCapacities.flush();
```
`link.flush();` does that in 1 line

guilpier-code replied:
you're right, I didn't call **flush()** on the link although I introduced this function in the link class

flomnes commented on src/solver/constraints-builder/cbuilder.cpp:
```diff
@@ -87,15 +87,12 @@ bool CBuilder::update(bool applyCheckBox)
-                                             Data::fhlMax,
-                                             (uint)8760,
-                                             Matrix<>::optImmediate))
+        // Try to open the file
+        if (not (*i)->ptr->loadDataFromCVSfile(Matrix<>::optImmediate))
```
You're de-referencing two pointers in a single instruction. This is dangerous since one of them could be `nullptr` or invalid.

---

It's unclear what `ptr` refers to, consider renaming this variable.

guilpier-code replied:
ok, **ptr** is a pointer to an object of type **AreaLink** (a study link actually).
For now, I do not rename that pointer, but I define a variable that makes this part of the code more clear.

flomnes commented on src/solver/constraints-builder/cbuilder.cpp:
```diff
@@ -106,124 +103,107 @@ bool CBuilder::update(bool applyCheckBox)
-        k->nImpedanceChanges = 0;
-        k->avgImpedance
-          = dataLink->entry[columnImpedance][0]; // or used a function created in matrix.h
+        (*i)->nImpedanceChanges = 0;
+        (*i)->avgImpedance = (*i)->ptr->parameters[columnImpedance][0];
```
2 de-reference in a single instruction. Consider using a variable for `*i`.

guilpier-code replied:
**(*i)** is now replaced with **linkInfo**

guilpier-code committed:
NTC multi TS - split link's data : changes due to code review

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/ntc-multi-ts--load-and-save' into feature/split-ntc-ts-container

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -83,12 +85,14 @@ void AreaLink::detach()
-    for (uint i = 0; i != data.height; ++i)
+    for (uint i = 0; i != HOURS_PER_YEAR; ++i)
     {
-        data[fhlNTCDirect][i] = 1.;
-        data[fhlNTCIndirect][i] = 1.;
+        directCapacities[0][i] = 1.;
+        indirectCapacities[0][i] = 1.;
```
Use `void Matrix<T, ReadWriteT>::fillColumn(uint x, const T& value)` instead.

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -289,6 +291,127 @@ static bool AreaLinksInternalLoadFromProperty(Study& study,
+    study.gotFatalError = true;
+}
+
+bool linkLoadTimeSeries_version_under_320(AreaLink& link, const AnyString& folder, Study& study)
```
Mark as `static`

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -289,6 +291,127 @@ static bool AreaLinksInternalLoadFromProperty(Study& study,
+    return ret;
+}
+
+bool linkLoadTimeSeries_version_320_to_630(AreaLink& link, const AnyString& folder)
```
Mark as `static`

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -289,6 +291,127 @@ static bool AreaLinksInternalLoadFromProperty(Study& study,
+            link.setPathToDataFile(buffer);
+            ret = link.loadDataFromCSVfile(Matrix<>::optFixedSize | Matrix<>::optImmediate) && ret;
+*/
+bool linkLoadTimeSeries_version_630_to_820(AreaLink& link, const AnyString& folder)
```
Mark as `static`

flomnes committed:
Merge pull request #459 from AntaresSimulatorTeam/develop

Merge develop
